src/functional_tests.rs

```diff
@@ -0,0 +1,117 @@
+
+use crate::{
+    actions::*,
+    state::*,
+    utilities::{
+        Candidate, Node, NodeState, ParsecVote, RelocatedInfo, Rpc, State, TestEvent, TryResult,
+    },
+};
+use rand::{self, Rng, SeedableRng};
+use rand_xorshift::XorShiftRng;
+use std::{env, iter, thread};
+use unwrap::unwrap;
+
+fn get_rng() -> XorShiftRng {
+    let env_var_name = "ROUTING_MODEL_SEED";
+    let seed = env::var(env_var_name)
+        .ok()
+        .map(|value| {
+            unwrap!(
+                value.parse::<u64>(),
+                "Env var 'ROUTING_MODEL_SEED={}' is not a valid u64.",
+                value
+            )
+        })
+        .unwrap_or_else(rand::random);
+    println!(
+        "To replay this '{}', set env var {}={}",
+        unwrap!(thread::current().name()),
+        env_var_name,
+        seed
+    );
+    XorShiftRng::seed_from_u64(seed)
+}
+
+#[test]
+fn relocate_adult_src() {
+    let adult_count = 2;
+    let elder_count = 3;
+    let mut rng = get_rng();
+    let mut nodes = iter::repeat_with(|| rng.gen())
+        .take(adult_count + elder_count)
+        .collect::<Vec<Node>>();
+    nodes.sort_unstable();
+    let (adults, elders) = nodes.split_at(adult_count);
+
+    let action = Action::new(
+        InnerAction::new_with_our_attributes(rng.gen())
+            .with_next_target_interval(rng.gen())
+            .extend_current_nodes_with(&NodeState::default_elder(), elders)
+            .extend_current_nodes_with(&NodeState::default(), adults),
+    );
+    let mut member_state = MemberState {
+        action,
+        ..Default::default()
+    };
+
+    let relocating_node = adults[0];
+    assert!(member_state
+        .action
+        .inner()
+        .our_current_nodes
+        .contains_key(&relocating_node.0.name));
+
+    let relocated_info = RelocatedInfo {
+        candidate: Candidate(relocating_node.0),
+        expected_age: relocating_node.0.age.increment_by_one(),
+        target_interval_centre: rng.gen(),
+        section_info: rng.gen(),
+    };
+
+    let required_events = [
+        TestEvent::SetWorkUnitEnoughToRelocate(relocating_node).to_event(),
+        ParsecVote::WorkUnitIncrement.to_event(),
+        ParsecVote::CheckRelocate.to_event(),
+        Rpc::RelocateResponse(relocated_info).to_event(),
+        ParsecVote::RelocateResponse(relocated_info).to_event(),
+    ];
+
+    let optional_random_events = [
+        ParsecVote::WorkUnitIncrement.to_event(),
+        ParsecVote::CheckRelocate.to_event(),
+    ];
+
+    for required_event in &required_events {
+        assert_eq!(TryResult::Handled, member_state.try_next(*required_event));
+        for optional_event in &optional_random_events {
+            if rng.gen() {
+                assert_eq!(TryResult::Handled, member_state.try_next(*optional_event));
+            }
+        }
+    }
+
+    let expected_state = unwrap!(member_state
+        .action
+        .inner()
+        .our_current_nodes
+        .get(&relocating_node.0.name)
```

Comment:
maybe add a get_node_state beside the set_node_state on Action?

Reply:
I've just realised that `set_node_state()` is only a private method of `InnerAction`, but I don't think that changes the fact that an `Action::node_state()` getter would be useful here?

Reply:
Actually, rather than unwrapping, I'll return an option since I can use that to check the relocated node is subsequently removed after handling the `ParsecVote::RelocatedInfo`.